crates/sui-transaction-builder/src/object.rs

```diff
@@ -0,0 +1,203 @@
+            mutable: None,
+        }
+    }
+
+    pub fn shared(id: ObjectId, initial_shared_version: u64, mutable: bool) -> Self {
+        Self {
+            id,
+            kind: Some(Kind::Shared),
+            version: None,
+            digest: None,
+            initial_shared_version: Some(initial_shared_version),
+            mutable: Some(mutable),
+        }
+    }
+
+    // Add partial information
+
+    // Kind
+    pub fn as_owned(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::ImmOrOwned),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+    // Redundant, but who ever liked saying "imm_or_owned"?
+    pub fn as_immutable(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::ImmOrOwned),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: Some(false),
+        }
+    }
+
+    pub fn as_receiving(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::Receiving),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    pub fn as_shared(self) -> Self {
+        Self {
+            id: self.id,
+            kind: Some(Kind::Shared),
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    // ObjectRef fields
+    pub fn versioned_at(self, version: u64) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: Some(version),
+            digest: self.digest,
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    pub fn with_digest(self, digest: ObjectDigest) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: self.version,
+            digest: Some(digest),
+            initial_shared_version: self.initial_shared_version,
+            mutable: self.mutable,
+        }
+    }
+
+    // Shared fields
+
+    // Initial shared version
+    pub fn shared_at(self, i: u64) -> Self {
+        Self {
+            id: self.id,
+            kind: self.kind,
+            version: self.version,
+            digest: self.digest,
+            initial_shared_version: Some(i),
+            mutable: self.mutable,
+        }
+    }
+
+    // Shared value mutability
+    pub fn by_val(self) -> Self {
+        self
+    }
+    pub fn by_ref(&self) -> &Self {
+        self
+    }
+    pub fn by_mut(&mut self) -> &mut Self {
+        self
+    }
```

Comment:
```suggestion
    // Shared value mutability
    pub fn by_val(self) -> Self {
        Self { mutable: Some(true), ..self }
    }
    pub fn by_ref(&self) -> &Self {
        Self { mutable: Some(false), ..self }
    }
    pub fn by_mut(&mut self) -> &mut Self {
        Self { mutable: Some(true), ..self }
    }
```

Reply:
I don't fully follow these functions. How would these be used?

Reply:
They are used to set the mutability of a shared object input, based on whether that input is being used by value, by reference, or by mutable reference.